x-pack/plugins/fleet/server/services/secrets.ts

```diff
@@ -102,47 +105,136 @@ export async function createSecrets(opts: {
+    soClient,
+    ids,
+  });
+
+  if (packagePoliciesUsingSecrets.length) {
+    packagePoliciesUsingSecrets.forEach(({ id, policyIds }) => {
+      logger.debug(
+        `Not deleting secret with id ${id} is still in use by package policies: ${policyIds.join(
+          ', '
+        )}`
+      );
+    });
+  }
+
+  const secretsToDelete = ids.filter((id) => {
+    return !packagePoliciesUsingSecrets.some((packagePolicy) => packagePolicy.id === id);
+  });
+
+  if (!secretsToDelete.length) {
+    return;
+  }
+
+  await _deleteSecrets({
+    esClient,
+    ids: secretsToDelete,
+  });
+}
+
+export async function findPackagePoliciesUsingSecrets(opts: {
+  soClient: SavedObjectsClientContract;
+  ids: string[];
+}): Promise<Array<{ id: string; policyIds: string[] }>> {
+  const { soClient, ids } = opts;
+  const packagePolicies = await packagePolicyService.list(soClient, {
+    kuery: `ingest-package-policies.secret_references.id: (${ids.join(' or ')})`,
+    perPage: SO_SEARCH_LIMIT,
+    page: 1,
+  });
+
+  if (!packagePolicies.total) {
+    return [];
+  }
+
+  // create a map of secret_references.id to package policy id
+  const packagePoliciesBySecretId = packagePolicies.items.reduce((acc, packagePolicy) => {
+    packagePolicy?.secret_references?.forEach((secretReference) => {
+      if (!acc[secretReference.id]) {
+        acc[secretReference.id] = [];
+      }
+      acc[secretReference.id].push(packagePolicy.id);
+    });
+    return acc;
+  }, {} as Record<string, string[]>);
+
+  const res = [];
+
+  for (const id of ids) {
+    if (packagePoliciesBySecretId[id]) {
+      res.push({
+        id,
+        policyIds: packagePoliciesBySecretId[id],
+      });
+    }
+  }
+
+  return res;
+}
+
+// use deleteSecretsIfNotReferenced
+export async function _deleteSecrets(opts: {
+  esClient: ElasticsearchClient;
+  ids: string[];
+}): Promise<void> {
+  const { esClient, ids } = opts;
+  const logger = appContextService.getLogger();
+  const body = ids.flatMap((id) => [
+    {
+      delete: { _index: getSecretsIndex(), _id: id },
+    },
+  ]);
+
+  let res: BulkResponse;
+
   try {
-    res = await esClient.delete({
-      index: getSecretsIndex(),
-      id,
+    res = await esClient.bulk({
+      body,
     });
 
-    auditLoggingService.writeCustomAuditLog({
-      message: `secret deleted: ${id}`,
-      event: {
-        action: 'secret_delete',
-        category: ['database'],
-        type: ['access'],
-        outcome: 'success',
-      },
+    const [errorItems, successItems] = partition(res.items, (a) => a.delete?.error);
+
+    successItems.forEach((item) => {
+      auditLoggingService.writeCustomAuditLog({
+        message: `secret deleted: ${item.delete!._id}`,
+        event: {
+          action: 'secret_delete',
+          category: ['database'],
+          type: ['access'],
+          outcome: 'success',
+        },
+      });
     });
+
+    if (errorItems.length) {
+      throw new Error(JSON.stringify(errorItems));
```

Comment:
Does this mean that the API call will return an error if can't delete an unused secret? I wonder if this should be a warning, since the purpose of the request is successful (to update package policy), and only the cleanup would fail in this case.

Reply:
resolved by https://github.com/elastic/kibana/pull/156806/commits/fd880314b67e5d2eae727436e68a5428d061c2a1